[Res][P2] Implement Response Parser Component

deerhunter4 opened the pull request:
response_parser.py created from part of the api_requester.py
dictionaries key names in response_parser.py were changed

deerhunter4 committed:
[Res][P2] Implement Response Parser Component
response_parser.py created from part of the api_requester.py
dictionaries key names response_parser.py were changed

KrzysiekJodlowski commented on response_parser.py:
```diff
@@ -0,0 +1,47 @@
+def parse_weatherapi_response(weatherapi_response):
```
I'm having second thoughts on the filename, what do you think about renaming it `api_response_parser` to make it more accurate?

deerhunter4 replied:
@KrzysiekJodlowski Done.

KrzysiekJodlowski commented on response_parser.py:
```diff
@@ -0,0 +1,47 @@
+    temp = []
+    rain = []
+
+    # extracting needed information from request response
```
Do you think we need this explanation here? I cannot see any of this kind in other methods.

deerhunter4 replied:
@KrzysiekJodlowski Done.

KrzysiekJodlowski commented on response_parser.py:
```diff
@@ -0,0 +1,47 @@
+        temp.append(item['temp_c'])
+        rain.append(item['precip_mm'])
+
+    latitude = response_dict['location']['lat']
+    longitude = response_dict['location']['lon']
```
I know we need these to be obtained here because the other parser methods depend on it. But I still don't like this, both:
- the fact that this parse method has more than one responsibility (it should care only about the weather like the other parse methods)
- and the fact that method call order depends on it (they should not, because e.g. what about the situation in which there's no weatherapi response?)

Ideally, we should have a separate method that would extract this data from an independent-from-weather-apis API endpoint. Do you think it's possible to make it that way? Have you already tried it and can tell how complex such an update would be?

deerhunter4 replied:
@KrzysiekJodlowski  
I created new function get_coordinates() that extracts longitude and latitude from weatherapi response. So now parse_weatherapi_response() parser only weather data like two other parser functions.
As for the second for the second issue, I agree with you. However, getting location coordinates from an independent source would require an additional API request and API key (registration, etc.). It also looks like that we would need extra parameters for some of the API services:
[https://openweathermap.org/api/geocoding-api](https://openweathermap.org/api/geocoding-api)
[https://www.geoapify.com/geocoding-api/](https://www.geoapify.com/geocoding-api/)
For now, I would leave it as it is. I wouldn't want to complicate it more for the user. In future, I might try to make an additional request to weatherapi.com just for coordinates. If I am correct there are language options, so it might even be possible to use Polish or Spanish names instead of English (for now it is impossible to get a forecast for Przemyśl).

KrzysiekJodlowski replied:
@deerhunter4 ok I got your point, and I agree that we can think about making it a proper way (with a separate API usage) later on, to not do more than is expected from this PR - please create a ticket in the backlog for that, I think that for phase 4 at max would be fine. 

And as long as it's cool that the request method can return more than one method call, It's a bit disturbing that it now [has two responsibilities](https://en.wikipedia.org/wiki/Single-responsibility_principle)/[these two methods are coupled together](https://en.wikipedia.org/wiki/Coupling_(computer_programming)). To mitigate this situation, because now it seems we start having [more complicated/sophisticated code](https://github.com/deerhunter4/Better_Weather/pull/14#pullrequestreview-2161600710), can I ask you to consider [the originally suggested approach](https://trello.com/c/Npa6CXXe) of having that logic blended into the `main`? I can see it more-or-less like this:

- instead of:
```python
# api_requester.py
return parse_weatherapi_response(weatherapi_response), get_coordinates(weatherapi_response)

# main.py
weatherapi_forecast, [latitude, longitude] = request_weatherapi(keys[WEATHERAPI_NAME], parameters)
```
- something like:
```python
# api_requester.py
return weatherapi_response

# main.py
weatherapi_response = request_weatherapi(keys[WEATHERAPI_NAME], parameters)
[latitude, longitude] = get_coordinates(weatherapi_response)
weatherapi_forecast = parse_weatherapi_response(weatherapi_response)
```

A similar approach, but less complicated, would apply to other API responses/parsing.  What do you think?

Apart from the above:
> I wouldn't want to complicate it more for the user

The user doesn't know the internals of your program, I think you meant yourself😉
> If I am correct there are language options (...)

That's a great idea! Though this is an extra thing - again, please create a ticket for that, this will be P5.

deerhunter4 replied:
@KrzysiekJodlowski  corrected according to guidelines.

KrzysiekJodlowski reviewed (commented):
@deerhunter4 overall looks fine:
- I like the fact you renamed variable names returned in those objects
- still, please try and address the comments I've left
- as it comes to the question you asked under [the Trello item](https://trello.com/c/Npa6CXXe/12-resp2-implement-response-parser-component) (that, btw, should be placed here where the discussion about the code changes take place) I agree to use this parser within the requester component. It's not so much about the standards but more about common sense - as the parser component is specific to the responses made by this requester, it should rather be used within. It could be another case e.g. if we'd have more complicated/sophisticated code, but here/for basic script/procedural structure like this, it makes sense.

deerhunter4 committed:
[Res][P2] enamed response_parser.py to api_response_parser.py

deerhunter4 committed:
[Res][P2] Update api_response_parser.py

deerhunter4 committed:
[Res][P2] get_coordinates() function created

get_coordinates() extracts longitude and latitude.
parse_weatherapi_response() parse only
weather data like two other parser functions.

deerhunter4 committed:
[Res][P2] Update api_requester.py and main.py
